peer/brontide.go

```diff
@@ -358,6 +365,84 @@ type Config struct {
 	Quit chan struct{}
 }
 
+// pingManager is a structure that is designed to manage the internal state
+// of the ping pong lifecycle with the remote peer. We assume there is only one
+// ping outstanding at once.
+type pingManager struct {
+	// pingTime is a rough estimate of the RTT (round-trip-time) between us
+	// and the connected peer.
+	// To be used atomically.
+	// TODO(roasbeef): also use a WMA or EMA?
+	pingTime atomic.Pointer[time.Duration]
+
+	// pingLastSend is the time when we sent our last ping message.
+	// To be used atomically.
+	pingLastSend atomic.Pointer[time.Time]
+
+	// outstandingPongSize is the current size of the requested pong
+	// payload.  This value can only validly range from [0,65531]. Any
+	// value < 0 is interpreted as if there is no outstanding ping message.
+	outstandingPongSize atomic.Int32
+
+	// timeoutTask is a pointer to the currently running timeout logic.
+	// this should be nil at all times that outstandingPongSize is < 0
+	timeoutTask atomic.Pointer[time.Timer]
```

Comment:
I already adjusted this code to make things as defensive/robust as possible. Unless you have a specific problem with this approach that you'd like to elaborate on, I'm not inclined to redesign it again.

Reply:
I think if things are grouped into a struct like it is now, then that struct should manage the atomicity of things within it. This pingManager as it stands is currently not atomic. It will also improve readability a ton and will make it easier to reason about concurrency in future if this struct expands. It is not a re-design i am suggesting, just a simple change. Here is a patch:

[pingmanager.patch](https://github.com/lightningnetwork/lnd/files/12534983/pingmanager.patch)